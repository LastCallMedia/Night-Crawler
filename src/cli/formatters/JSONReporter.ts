import Reporter from './Reporter';
import { TestResult, TestResultMap } from '../../testing/TestContext';
import { writeFile } from 'fs';
import { promisify } from 'util';

const writeFileP = promisify(writeFile);

function mapToObj<T = unknown>(map: Map<string, T>): { [k: string]: T } {
  const obj = Object.create(null);
  for (const [k, v] of map) {
    obj[k] = v;
  }
  return obj;
}

export default class JSONReporter implements Reporter {
  path: string;
  collected: { url: string; result: { [k: string]: TestResult } }[];
  constructor(path: string) {
    this.path = path;
    this.collected = [];
  }
  async start(): Promise<void> {
    // no-op.
  }
  report(url: string, result: TestResultMap): void {
    this.collected.push({ url, result: mapToObj<TestResult>(result) });
  }
  async stop(): Promise<void> {
    await writeFileP(this.path, JSON.stringify(this.collected, null, 4));
  }
}
