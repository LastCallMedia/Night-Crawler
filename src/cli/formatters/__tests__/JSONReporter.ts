import { file } from 'tmp-promise';
import { promises as fs } from 'fs';
import { TestResultMap, TestResult } from '../../../testing/TestContext';
import JSONReporter from '../JSONReporter';

function r(obj: { [k: string]: TestResult }): TestResultMap {
  return new Map(Object.entries(obj));
}

describe('JSON Reporter', function() {
  const pass = r({
    pass: { pass: true }
  });
  const fail = r({
    fail: { pass: false, message: 'There was an error.' }
  });
  const mix = r({
    pass: { pass: true },
    fail: { pass: false, message: 'there was an error.' }
  });

  it('Should write a json file containing results', async function() {
    const { path, cleanup } = await file();
    const reporter = new JSONReporter(path);
    reporter.report('passing', pass);
    reporter.report('failing', fail);
    reporter.report('mixed', mix);
    await expect(reporter.stop()).resolves.toBeUndefined();
    await expect(fs.readFile(path, 'utf8')).resolves.toMatchSnapshot();
    await cleanup();
  });
});
