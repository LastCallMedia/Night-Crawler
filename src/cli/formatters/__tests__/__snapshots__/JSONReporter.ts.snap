// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON Reporter Should write a json file containing results 1`] = `
"[
    {
        \\"url\\": \\"passing\\",
        \\"result\\": {
            \\"pass\\": {
                \\"pass\\": true
            }
        }
    },
    {
        \\"url\\": \\"failing\\",
        \\"result\\": {
            \\"fail\\": {
                \\"pass\\": false,
                \\"message\\": \\"There was an error.\\"
            }
        }
    },
    {
        \\"url\\": \\"mixed\\",
        \\"result\\": {
            \\"pass\\": {
                \\"pass\\": true
            },
            \\"fail\\": {
                \\"pass\\": false,
                \\"message\\": \\"there was an error.\\"
            }
        }
    }
]"
`;
